packages/aws-cdk/lib/api/hotswap/lambda-functions.ts

```diff
@@ -46,28 +46,22 @@ export function isHotswappableLambdaFunctionChange(
 }
 
 /**
- * Returns `true` if the change is not for a AWS::Lambda::Function,
+ * Returns `ChangeHotswapImpact.IRRELEVANT` if the change is not for a AWS::Lambda::Function,
  * but doesn't prevent short-circuiting
  * (like a change to CDKMetadata resource),
- * `false` if the change is to a AWS::Lambda::Function,
+ * `ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT` if the change is to a AWS::Lambda::Function,
  * but not only to its Code property,
  * or a LambdaFunctionCode if the change is to a AWS::Lambda::Function,
  * and only affects its Code property.
  */
 function isLambdaFunctionCodeOnlyChange(
   change: cfn_diff.ResourceDifference, assetParamsWithEnv: { [key: string]: string },
 ): LambdaFunctionCode | ChangeHotswapImpact {
-  if (!change.newValue) {
-    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
-  }
-  const newResourceType = change.newValue.Type;
-  // Ignore Metadata changes
-  if (newResourceType === 'AWS::CDK::Metadata') {
-    return ChangeHotswapImpact.IRRELEVANT;
-  }
-  // The only other resource change we should see is a Lambda function
+  // if we see a different resource type, it will be caught by isNonHotswappableResourceChange()
+  // this also ignores Metadata changes
+  const newResourceType = change.newValue?.Type;
   if (newResourceType !== 'AWS::Lambda::Function') {
-    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+    return ChangeHotswapImpact.IRRELEVANT;
   }
   if (change.oldValue?.Type == null) {
```

Comment:
Feels like this check should also be moved to `isNonHotswappableResourceChange()`.